### PR TITLE
Add docstrings to `patroni.ctl`

### PR DESCRIPTION
**Pull request by barthisrael**
References: PAT-90.

### CODE REVIEW

File: patroni/ctl.py [hunk 1]
```diff
@@ -723,13 +1144,36 @@ def reinit(obj: Dict[str, Any], cluster_name: str, group: Optional[int],
...
+
+    .. note::
+        If not able to perform the operation through the REST API, fall back to using the old manual failover approach,
+        directly through the DCS.
+
+    :param obj: Patroni configuration.
+    :param action: action to be taken -- ``failover`` or ``switchover``.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter Citus group within we should perform a failover or switchover. If ``None`` consider members
+        from the coordinator group.
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/ctl.py:1160]: In theory it should be possible to find the group from the provided `leader` or `candidate`, but it is not yet implemented.

File: patroni/ctl.py [hunk 2]
```diff
@@ -846,6 +1290,24 @@ def _do_failover_or_switchover(obj: Dict[str, Any], action: str, cluster_name: s
...
+    .. note::
+        If *leader* is given perform a switchover instead of a failover.
+
+    .. seealso::
+        Refer to :func:`_do_failover_or_switchover` for details.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter Citus group within we should perform a failover or switchover. If ``None`` consider members of
+        the coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:1306]: The same as https://github.com/zalando/patroni/pull/2687/files#diff-9dada8f509128404d7027246ac9622d2484fc77758c6a4af9dc7cc8c081d9fffR1160

File: patroni/ctl.py [hunk 3]
```diff
@@ -908,6 +1441,36 @@ def get_cluster_service_info(cluster: Dict[str, Any]) -> List[str]:
...
+
+    Information is printed to console through :func:`print_output`, and contains:
+
+        * ``Cluster``: name of the Patroni cluster, as per ``scope`` configuration;
+        * ``Member``: name of the Patroni node, as per ``name`` configuration;
+        * ``Host``: hostname (or IP) and port, as per ``postgresql.listen`` configuration;
+        * ``Role``: ``Leader``, ``Standby Leader``, ``Sync Standby`` or ``Replica``;
+        * ``State``: ``stopping``, ``stopped``, ``stop failed``, ``crashed``, ``running``, ``starting``,
+          ``start failed``, ``restarting``, ``restart failed``, ``initializing new cluster``, ``initdb failed``,
+          ``running custom bootstrap script``, ``custom bootstrap failed``, or ``creating replica``;
```
Comment (CyberDem0n) [inline patroni/ctl.py:1454]:
```suggestion
        * ``State``: ``stopping``, ``stopped``, ``stop failed``, ``crashed``, ``running``, ``starting``,
          ``start failed``, ``restarting``, ``restart failed``, ``initializing new cluster``, ``initdb failed``,
          ``running custom bootstrap script``, ``custom bootstrap failed``, or ``creating replica``, and so on;
```

There could be more :)

File: patroni/ctl.py [hunk 4]
```diff
@@ -1024,6 +1620,18 @@ def timestamp(precision: int = 6) -> str:
 @click.pass_obj
 def flush(obj: Dict[str, Any], cluster_name: str, group: Optional[int],
           member_names: List[str], force: bool, role: str, target: str) -> None:
+    """Process ``flush`` command of ``patronictl`` utility.
+
+    Discard scheduled restart or switchover events.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group we should flush an event. If ``None`` assume coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:1629]: The same

File: patroni/ctl.py [hunk 5]
```diff
@@ -1081,6 +1696,18 @@ def wait_until_pause_is_applied(dcs: AbstractDCS, paused: bool, old_cluster: Clu
...
+    """Toggle the ``pause`` state in the cluster members.
+
+    :param config: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group we should toggle the pause state of. If ``None`` assume coordinator group.
+    :param paused: the desired state for ``pause`` in all nodes.
+    :param wait: ``True`` if it should block until the operation is finished or ``false`` for returning immediately.
+
+    :raises PatroniCtlException: if
+        * ``pause`` state is already *paused*; or
```
Comment (CyberDem0n) [inline patroni/ctl.py:1708]:
```suggestion
        * ``pause`` state is already *paused*/*unpaused*; or
```

File: patroni/ctl.py [hunk 6]
```diff
@@ -1114,6 +1741,15 @@ def toggle_pause(config: Dict[str, Any], cluster_name: str, group: Optional[int]
 @click.pass_obj
 @click.option('--wait', help='Wait until pause is applied on all nodes', is_flag=True)
 def pause(obj: Dict[str, Any], cluster_name: str, group: Optional[int], wait: bool) -> None:
+    """Process ``pause`` command of ``patronictl`` utility.
+
+    Put the cluster in maintenance mode.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group we should pause. If ``None`` assume coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:1750]: The same

File: patroni/ctl.py [hunk 7]
```diff
@@ -1123,17 +1759,27 @@ def pause(obj: Dict[str, Any], cluster_name: str, group: Optional[int], wait: bo
 @click.option('--wait', help='Wait until pause is cleared on all nodes', is_flag=True)
 @click.pass_obj
 def resume(obj: Dict[str, Any], cluster_name: str, group: Optional[int], wait: bool) -> None:
+    """Process ``unpause`` command of ``patronictl`` utility.
+
+    Put the cluster out of maintenance mode.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group we should unpause. If ``None`` assume coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:1768]: The same

File: patroni/ctl.py [hunk 8]
```diff
@@ -1310,6 +2001,26 @@ def invoke_editor(before_editing: str, cluster_name: str) -> Tuple[str, Dict[str
 def edit_config(obj: Dict[str, Any], cluster_name: str, group: Optional[int],
                 force: bool, quiet: bool, kvpairs: List[str], pgkvpairs: List[str],
                 apply_filename: Optional[str], replace_filename: Optional[str]) -> None:
+    """Process ``edit-config`` command of ``patronictl`` utility.
+
+    Update or replace Patroni configuration in the DCS.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group configuration we should edit. If ``None``  edit from coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:2010]: The same

File: patroni/ctl.py [hunk 9]
```diff
@@ -1358,6 +2069,14 @@ def edit_config(obj: Dict[str, Any], cluster_name: str, group: Optional[int],
 @option_default_citus_group
 @click.pass_obj
 def show_config(obj: Dict[str, Any], cluster_name: str, group: Optional[int]) -> None:
+    """Process ``show-config`` command of ``patronictl`` utility.
+
+    Show Patroni configuration stored in the DCS.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group configuration we should show. If ``None`` show from coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:2078]: The same

File: patroni/ctl.py [hunk 10]
```diff
@@ -1369,6 +2088,18 @@ def show_config(obj: Dict[str, Any], cluster_name: str, group: Optional[int]) ->
...
+    """Process ``version`` command of ``patronictl`` utility.
+
+    Show version of:
+        * ``patronictl`` on invoker;
+        * ``patroni`` on all members of the cluster;
+        * ``PostgreSQL`` on all members of the cluster.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group we should get members from. If ``None`` assume coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:2100]: The same

File: patroni/ctl.py [hunk 11]
```diff
@@ -1396,6 +2127,22 @@ def version(obj: Dict[str, Any], cluster_name: str, group: Optional[int], member
...
+    Information is printed to console through :func:`print_output`, and contains:
+        * ``TL``: Postgres timeline when the event occurred;
+        * ``LSN``: Postgres LSN, in bytes, when the event occurred;
+        * ``Reason``: the reason that motivated the event, if any;
+        * ``Timestamp``: timestamp when the event occurred;
+        * ``New Leader``: the Postgres node that was promoted during the event.
+
+    :param obj: Patroni configuration.
+    :param cluster_name: name of the Patroni cluster.
+    :param group: filter which Citus group we should get events from. If ``None`` assume coordinator group.
```
Comment (CyberDem0n) [inline patroni/ctl.py:2143]: The same

**Comment by CyberDem0n** [comment]
Now only a few uncovered lines in the validator.py left: https://coveralls.io/builds/60313576/source?filename=patroni%2Fpostgresql%2Fvalidator.py#L249 :)

File: patroni/ctl.py (hunk 5 above)
Comment (barthisrael) [reply patroni/ctl.py:1707]:
```suggestion
        * Cluster ``is_paused`` state is already the same as *paused*; or
```

With this one I meant that `is_paused` attribute is already the same as `paused` argument.
The `*` is usually used to reference an argument, so I guess we should not use *unpaused*.
What do you think of the above instead?

File: patroni/ctl.py [hunk 12]
```diff
@@ -304,12 +466,29 @@ def get_all_members(obj: Dict[str, Any], cluster: Cluster,
 
 def get_any_member(obj: Dict[str, Any], cluster: Cluster, group: Optional[int],
                    role: str = 'leader', member: Optional[str] = None) -> Optional[Member]:
+    """Get the first found cluster member that has the given *role*.
+
+    :param obj: the Patroni configuration.
+    :param cluster: the Patroni cluster.
+    :param group: filter which Citus group we should get members from. If ``None`` get from all groups.
+    :param role: role to filter members. See :func:`get_all_members` for available options.
+    :param member: if specified, then besides having the given *role*, the cluster member's name should be *member*.
```
[the remark this answers is not in the document]
Comment (barthisrael) [reply patroni/ctl.py:484]: I think that makes sense, yes. I'll submit a commit with that change.

**Review by CyberDem0n — COMMENTED** [review]
Most of the commands (restart/reinit/pause/resume/show-config/edit-config and similar) require the group argument and work only for this specific Citus group.
If not specified in the command line the group might be taken from the config file.
If it is also missing in the config file we assume that this is just a normal Patroni cluster (not Citus).

**Comment by barthisrael** [comment]
> Most of the commands (restart/reinit/pause/resume/show-config/edit-config and similar) require the group argument and work only for this specific Citus group.
> If not specified in the command line the group might be taken from the config file.
> If it is also missing in the config file we assume that this is just a normal Patroni cluster (not Citus).

Thanks! I added that as a note to the module, and made references from `group` param of the `patronictl` command to that note. What do you think of thew new version?

File: patroni/ctl.py [hunk 13]
```diff
@@ -303,13 +481,40 @@ def get_all_members(obj: Dict[str, Any], cluster: Cluster,
...
+
+    :returns: the first found cluster member that has the given *role*.
+
+    :raises:
+        :class:`PatroniCtlException`: if * both *role* and *member* are provided.
```
Comment (hughcapet) [inline patroni/ctl.py:496]:
```suggestion
        :class:`PatroniCtlException`: if both *role* and *member* are provided.
```

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**